Implementing DB Auth & Sessions

Quantity-4 opened the pull request:
Session, DB, and basic login functions added.

Needs bugfixing, login 'action' isnt called in the appropriate file.

Quantity-4 committed:
Implementing DB Auth & Sessions

Session, DB, and basic login functions added.

Needs bugfixing, login 'action' isnt called in the appropriate file.

Quantity-4 committed:
Implementing DB Auth & Sessions

Session, DB, and basic login functions added.

Needs bugfixing, login 'action' isnt called in the appropriate file.

Quantity-4 committed:
Create session, connect with database and login

This is the first working version of the login system, will need to clean it up, and add comments as well as delete unused dependencies before a pull.

Co-Authored-By: Jack <jack@mawersoft.co.uk>

Quantity-4 committed:
Create session, connect with database and login

This is the first working version of the login system, will need to clean it up, and add comments as well as delete unused dependencies before a pull.

Co-Authored-By: Jack <jack@mawersoft.co.uk>

Quantity-4 committed:
Clear login/error messages. Fixed Comments

Commenting was the main focus for better documentation.

Login messages are now also clearer - environment variable changes too.

Quantity-4 committed:
Clear login/error messages. Fixed Comments

Commenting was the main focus for better documentation.

Login messages are now also clearer - environment variable changes too.

Quantity-4 committed:
Update +page.server.js

Quantity-4 committed:
Correct cookies parameter documentation

jackmawer commented on .gitignore:
```diff
@@ -13,8 +13,9 @@ node_modules/
 
 # Env
 .env
-.env.*
```
These .gitignore rules for .env are redundant - ideally just put all of your development env vars in a single .env file for simplicity, no need for .env.database or whatever.

jackmawer commented on package.json:
```diff
@@ -25,6 +25,11 @@
+    "busboy": "^1.6.0",
+    "form-data": "^4.0.0",
+    "form-data-parser": "^0.0.3",
+    "formidable": "^2.1.1",
```
What are these form data parser libraries for, where are they used, and why are there four of them?

Quantity-4 replied:
Left over dependencies in order to attempt different ways of form-data parsing as I had some issues reading data in. I'll cleanup unused ones.

jackmawer commented on src/lib/server/db.js:
```diff
@@ -1,11 +1,287 @@
-const connectionString = process.env.VITE_PG_CONNECTION_STRING;
+
+// Load environment variables from the .env file
+dotenv.config();
```
Need to double check but this may be unneccessary due to how dotenv is setup in this environment.

Quantity-4 replied:
Safe to delete, if ENV variable can be accessed directly from .env file [request me a sample via DMs of the env file]

jackmawer commented on src/lib/server/db.js:
```diff
@@ -1,11 +1,287 @@
+dotenv.config();
+
+// Connection string for the PostgreSQL database
+const connectionString = process.env.VITE_PG_CONNECTION_STRING_LOCAL;
```
This is a weird env var name, why does it reference vite?

Quantity-4 replied:
Just a temp name, variable can be renamed, was creating it when initially setting up the vite environment [related to SSR] so thought it would be appropriate.

jackmawer commented on src/lib/server/db.js:
```diff
@@ -1,11 +1,287 @@
+ * @returns {string} A 32-character long hexadecimal string representing the session ID.
+ */
+function generateSessionId() {
+  return crypto.randomBytes(16).toString("hex");
```
`crypto` has a built-in function for generating UUIDs - try `crypto.randomUUID()`

Quantity-4 replied:
Updated method

jackmawer commented on src/lib/server/db.js:
```diff
@@ -1,11 +1,287 @@
+ * @returns {Promise<string>} The session ID of the created session.
+ * @throws {Error} If an error occurs while creating the session.
+ */
+export async function createSession(user) {
```
Not 100% about the function naming here, but I don't have a good suggestion so if you can't think of anything that would clarify what the actual function does then ignore this

jackmawer commented on src/lib/server/db.js:
```diff
@@ -1,11 +1,287 @@
+  const expirationHours = 24;
+  const currentDate = new Date();
+  currentDate.setHours(currentDate.getHours() + expirationHours);
+  return currentDate;
```
Naming becomes dodgy here because currentDate stops being the current date by line 134 - maybe worth rewriting this to be a bit clearer?

jackmawer commented on src/lib/server/db.js:
```diff
@@ -18,5 +294,6 @@ export async function testConnection() {
   }
 }
 
+// TODO: Debugging functions
```
duplicate heading

jackmawer commented on src/lib/server/db.js:
```diff
@@ -18,5 +294,6 @@ export async function testConnection() {
+// TODO: Debugging functions
 //TODO: debugging purposes only, this file should not normally be run directly
-testConnection();
+//testConnection();
```
All of this code is commented, remove it if you're not using it, we have git for version history

jackmawer commented on src/lib/server/session.js:
```diff
@@ -0,0 +1,53 @@
+ */
+export async function getUserFromCookie(cookieValue) {
+  // Get the user from the session cookie
+  return await getUserFromSession(cookieValue);
```
Isn't this function redundant? If there's a good reason to keep it around for something you have planned that seems fine to me, but why are we duplicating a function?

jackmawer commented on src/routes/auth/login/+page.server.js:
```diff
@@ -1,29 +1,66 @@
+ * @param {Object} cookies An object returning parameters from the request (params) and an object that allows access to cookies (cookies)
  * @returns {Promise<Object | null>} A user object, or null if not logged in.
- * @type {import('./$types').PageServerLoad}
+ * @type {import("./$types").PageServerLoad}
```
Why the quotes style change?

jackmawer commented on src/routes/auth/login/+page.server.js:
```diff
@@ -1,29 +1,66 @@
+    const body = await request.arrayBuffer();
+    const decodedBody = new TextDecoder().decode(body);
+    const formData = querystring.parse(decodedBody);
+    const { username, password } = formData;
```
This is needlessly overcomplicated! Use [Svelte's built-in form data parser](https://kit.svelte.dev/docs/form-actions#anatomy-of-an-action).
```js
const formData = await request.formData();
const username = data.get('username');
const password = data.get('password');
```

jackmawer commented on src/routes/auth/login/+page.server.js:
```diff
@@ -1,29 +1,66 @@
+          "Content-Type": "application/json"
+        }
+      };
+    }
```
We are not returning a response object, just data. For errors, use `fail()` - see https://kit.svelte.dev/docs/form-actions#anatomy-of-an-action-validation-errors

jackmawer commented on src/routes/auth/login/+page.server.js:
```diff
@@ -1,29 +1,66 @@
+        "Content-Type": "application/json",
+        "Set-Cookie": `sessionid=${sessionId}; Path=/; HttpOnly`
+      }
+    };
```
Again, we are not returning a response object, just data - see https://kit.svelte.dev/docs/form-actions#anatomy-of-an-action

jackmawer commented on src/routes/auth/login/+page.svelte:
```diff
@@ -4,34 +4,113 @@
     @see https://kit.svelte.dev/docs/form-actions
     */
-  import { redirect } from '@sveltejs/kit';
+  import { redirect } from "@sveltejs/kit";
```
Weird quotes again!

jackmawer commented on src/routes/auth/login/+page.svelte:
```diff
@@ -4,34 +4,113 @@
-    throw redirect(307, '/');
+    console.log("Logged in already");
+    throw redirect(302, "/posts");
+    window.location.href = "/posts";
```
My mistake, `redirect()` is for server code only. The correct function to use is `goto()` - https://kit.svelte.dev/docs/modules#$app-navigation-goto

Quantity-4 replied:
go-to causes errors for some reason, in the rendering of the page, I'll look more into this

jackmawer commented on src/routes/auth/login/+page.svelte:
```diff
@@ -4,34 +4,113 @@
+      errorMessage = error.body;
+    }
+
+  }
```
This whole function is unnecessary. Form submissions are handled by Svelte, you do not need to manually POST data - I'd suggest reading over the whole of https://kit.svelte.dev/docs/form-actions

Quantity-4 replied:
Using svelte's implementations kept giving me issues with parsing the data back to the server in an appropriate format - will look into it further

jackmawer commented on src/routes/auth/login/+page.svelte:
```diff
@@ -4,34 +4,113 @@
+
+  // Variables for username/password field
+  let usernameInput;
+  let passwordInput;
```
These two variables are unnecessary, Svelte handles form submissions with named inputs already.

jackmawer commented on src/routes/auth/login/+page.svelte:
```diff
@@ -4,34 +4,113 @@
+        <input name="password" type="password" bind:this="{passwordInput}" />
+      </label>
+      <button type="submit">Log in</button>
+    </form>
```
See above - do not override Svelte's form handling!

jackmawer commented on src/routes/posts/[slug]/+page.server.js:
```diff
@@ -1,5 +1,4 @@
-import { error } from "@sveltejs/kit";
-
+import { error} from "@sveltejs/kit";
```
Personal preference, I like a space between imports and our code.

jackmawer reviewed (changes requested):
A fantastic first push at getting the database side of things going, but a few misunderstandings especially regarding Svelte mechanics.

jackmawer commented on src/lib/server/db.js:
```diff
@@ -1,11 +1,287 @@
+/**
+ * Creates a new session for the given user in the sessions table.
+ *
+ * @param {number} user The user ID to create the session for.
```
Is this meant to be a number or a users object? session.js passes an object to this parameter!

Quantity-4 replied:
Object, need to make changes so its more legible

Quantity-4 committed:
Modified methods and Fixed project.json

Quantity-4 committed:
Merge branch 'db-auth-implement' of https://github.com/dss-ug08/dss-blog into db-auth-implement

Quantity-4 committed:
Update db.js

Quantity-4 committed:
Update package.json

jackmawer committed:
Add csrb bypass for development purposes

- GitHub codespaces don't work with Svelte if the default CSRF protection is enabled
- TODO: This MUST be re-enabled for production

Quantity-4 committed:
Able to register a user

Uses the same data parsing methods as in Login.

jackmawer commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Implement User Registration](https://trello.com/c/IKOfUCvq/4-implement-user-registration)

jackmawer commented:
This isn't perfect but I'm going to merge it in this case as the issues will be fixed by future PRs.
